desc/coils.py

```diff
@@ -1328,6 +1481,756 @@ def _check_type(coil0, coil):
+
+        names = ["p_frame", "q_frame", "curv1_frame", "curv2_frame", "x"]
+
+        if params is not None:
+            B_fb_params = params | B_fb_params
+        else:
+            B_fb_params = get_params(names, obj=self, basis="rpz") | B_fb_params
+
+        if transforms is None:
+            transforms = get_transforms(
+                names,
+                obj=self,
+                grid=finite_build_grid,
+                jitable=True,
+            )
+
+        compute_kwargs = {}
+        # pass the basis_in arg to compute_fun if necessary
+        try:
+            compute_kwargs["basis_in"] = self.basis
+        except KeyError:
+            pass
+
+        B_0_fb = compute_fun(
+            self,
+            "B_0_fb",
+            transforms=transforms,
+            params=B_fb_params,
+            profiles={},
+            **compute_kwargs,
+        )["B_0_fb"]
+        B_kappa_fb = compute_fun(
+            self,
+            "B_kappa_fb",
+            transforms=transforms,
+            params=B_fb_params,
+            profiles={},
+            **compute_kwargs,
+        )["B_kappa_fb"]
+
+        # get position of field measurement points in lab frame
+        x_fb = compute_fun(
+            self,
+            "x_fb",
+            transforms=transforms,
+            params=B_fb_params,
+            profiles={},
+            **compute_kwargs,
+        )["x_fb"]
+
+        x_fb = rpz2xyz(x_fb)
+
+        x_center = xyz2rpz(data["x"])
+        x_center = finite_build_grid.expand(x_center, "zeta")
+        phi = x_center[:, 1]
+        # the vectors for the B components are implicitly anchored to the CENTERLINE
+
+        B_0_fb = rpz2xyz_vec(B_0_fb, phi=phi)
+        B_kappa_fb = rpz2xyz_vec(B_kappa_fb, phi=phi)
+
+        # expand the vector series to include the cross section dimensions
+        B_b_fb = finite_build_grid.expand(B_b_fb, "zeta")
+        B_reg_fb = finite_build_grid.expand(B_reg_fb, "zeta")
+
+        B_self = B_0_fb + B_kappa_fb + B_b_fb + B_reg_fb
+
+        if coil_frame:  # reproject the field to the t,p,q frame
+            B_self = self.project_coil_frame(
+                B_self, finite_build_grid, centerline_grid, params
+            )
+
+        # find mask for the grid axis to prevent downstream singularities with
+        # biot savart
+        centerline_mask = jnp.zeros(len(finite_build_grid.nodes), dtype=bool)
+        centerline_mask = centerline_mask.at[
+            (L + 1) * (2 * M + 1) // 2 :: (L + 1) * (2 * M + 1)
+        ].set(True)
+
+        return B_self, x_fb, centerline_mask
+
+    def compute_self_field_circ(
+        self,
+        finite_build_grid,
+        centerline_grid,
+        coil_frame=False,
+        params=None,
+        transforms=None,
+    ):
+        return NotImplementedError(
+            "Circular cross section self field not implemented yet"
+        )
+
+    def project_coil_frame(
```

Comment:
Is this method needed? I feel like comptuting projections is pretty straightforward?